_source/_posts/2021-11-16-k8s-security-best-practices.md

```diff
@@ -0,0 +1,148 @@
+  - kind: Group
+    name: k8s-restricted-users
+```
+
+## 2. Secure the API server with OpenID Connect
+
+Kubernetes supports many authentication mechanisms. Some of the most common are:
+
+- Client certificates
+- Basic authentication
+- Tokens (Service account tokens, Bearer tokens, and so on)
+- [OpenID Connect (OIDC)](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens)
+- Proxy
+
+Out of all these authentication mechanisms, OIDC is the most secure and scalable solution. It is ideal for clusters accessed by large teams as it provides a single sign-on solution for all users and makes it easy to onboard and offboard users. It is also way more secure than other mechanisms as you don't have to store any sensitive information on a user's computer, like client secrets and passwords. You can also use features like [MFA](https://en.wikipedia.org/wiki/Multi-factor_authentication) and Yubikey if supported by your OIDC provider.
+
+{% img blog/k8s-api-server-oidc/kube-login-oidc.jpg alt:"K8s OIDC flow" width:"800" %}{: .center-image }
+
+OIDC combined with RBAC becomes necessary as more and more people start accessing the cluster. It becomes essential to create groups and roles and provide limited access to specific groups. You can read more about this in my previous post [How to Secure Your Kubernetes Cluster with OpenID Connect and RBAC](https://developer.okta.com/blog/2021/11/08/k8s-api-server-oidc).
+
+## 3. Use Secrets for all sensitive data with appropriate access
+
+This one should be a no-brainer. Kubernetes has a [Secret](https://kubernetes.io/docs/concepts/configuration/secret) resource that can be used to store sensitive data. This is a great way to store passwords, keys, and other sensitive data. Secrets can be used for storing string data, docker config, certificates, tokens, files, and so on. Secrets can be mounted as data volumes or exposed as environment variables to be used in containers. Secrets can be plain text or encoded, but please don't be that person who uses plain text secrets.
+
+Secrets are flexible and native to Kubernetes, so there is no reason for you not to use them. Also, make sure to implement proper RBAC for secrets so that not everyone has access to them.
+
+## 4. Keep Kubernetes version up to date
+
+Like any other software, Kubernetes also has bugs and issues. And from time to time, there might be a high severity bug that calls for a [CVE](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). Hence, it's an excellent idea to keep the Kubernetes version up-to-date on the server and the CLI client. If you are using a managed PaaS, it should be pretty easy to upgrade, and for on-prem installations, there are tools like [kOps](https://kops.sigs.k8s.io/), [kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/), and so on, that makes it easy to upgrade clusters.
```

Comment:
optional: could be helpful to include https://kubernetes.io/docs/reference/issues-security/security/ in here for those who might not think to search it to find the disclosure list